### PR TITLE
Add equivalents of C's <ctype.h> functions to AsciiExt.

### PR DESCRIPTION
**Pull request by zackw**
* `is_ascii_alphabetic`
 * `is_ascii_uppercase`
 * `is_ascii_lowercase`
 * `is_ascii_alphanumeric`
 * `is_ascii_digit`
 * `is_ascii_hexdigit`
 * `is_ascii_punctuation`
 * `is_ascii_graphic`
 * `is_ascii_whitespace`
 * `is_ascii_control`

This addresses issue #39658.

Lightly tested on x86-64-linux.  tidy complains about the URLs in the documentation making lines too long, I don't know what to do about that.

### CODE REVIEW

**Comment by sfackler** [comment]
cc @rust-lang/libs 

I've always been a bit annoyed this is a separate trait at all. Might be worth inlining these while we're at it.

**Comment by zackw** [comment]
> I've always been a bit annoyed this is a separate trait at all. Might be worth inlining these while we're at it.

I don't disagree, but can that be done compatibly? There's already a compatibility kludge (the `unimplemented!`s) in here...

**Comment by sfackler** [comment]
The idea would be that we deprecate this entire module and just move to inherent methods.

**Comment by aturon** [comment]
@sfackler 

> I've always been a bit annoyed this is a separate trait at all. Might be worth inlining these while we're at it.

Interesting -- I don't think there's a good rationale for them being extension methods right now. (Maybe there was at some point, pre-DST). Given that all the methods include `ascii` in their name, this seems reasonable to me.

I'm also :+1: on this PR.

**Comment by zackw** [comment]
I don't mind doing the work to turn `AsciiExt` into inherent methods, but I don't understand how libstd, libcore, and libcollections are organized well enough to know where to _put_ the inherent methods.

**Comment by aturon** [comment]
@japaric, any chance you'd be willing to help @zackw here, given that it may require some more inherent method magic around primitive types?

**Comment by alexcrichton** [comment]
Looks good to me as well!

I believe the intention of the trait was to add the method to `[u8]` and `u8` as well, as opposed to just `char` and `str`.

I'd be tempted to punt the migration to inherent methods (if at all) to a future PR though and go ahead and land this. Thoughts?

**Comment by alexcrichton** [comment]
@zackw it also looks like tidy may be failing on travis?

**Comment by aturon** [comment]
> I'd be tempted to punt the migration to inherent methods (if at all) to a future PR though and go ahead and land this. Thoughts?

Yeah, that probably makes sense. I do think it'd be a nice change, but it's orthogonal.

**Comment by zackw** [comment]
@alexcrichton Yeah, that's because the added documentation contains two very long URLs.  I have just pushed a change to squeeze those lines in under the limit (by shortening the `[foo]:` tag for them).

**Comment by alexcrichton** [comment]
Shame on you for writing thorough documentation! 

:)

**Comment by zackw** [comment]
Okay, now the problem is that the doc tests for the new unstable feature aren't properly annotated to say that they use the new unstable feature. 😝  And the reason I didn't notice this before is that I don't know how to run stdlib doc tests locally.  :man_facepalming: 

Can one of you please tell me how to do that?

**Comment by alexcrichton** [comment]
Oh you can just prefix the tests with `#![feature(...)]` to allow them to pass. There should be a few other examples of this throughout the standard library as well

**Comment by zackw** [comment]
> Oh you can just prefix the tests with `#![feature(...)]` to allow them to pass

That part I know.  What I don't know is how to _run the doc tests locally_, so that I can verify that I did it right before pushing.

**Comment by alexcrichton** [comment]
Oh for that you can use `./x.py src/libstd --test-args $name`

**Comment by aturon** [comment]
We discussed this is the libs team meeting, and we're all in favor of merging as-is. Feel free to open a separate PR to move to inherent methods as well!

r=me once the travis failure is addressed.

**Comment by zackw** [comment]
@alexcrichton Oh, I see now.  `x.py test src/libstd` *does* run the doctests, but only If all of the `#[test]` tests succeed.  And my computer can't do IPv6, which makes the `#[test]` tests always have some failures, so I thought the doctests were under a separate x.py invocation.  Thanks for the hint.

**Comment by japaric** [comment]
@aturon 

> @japaric, any chance you'd be willing to help @zackw here, given that it may require some more inherent method magic around primitive types?

Sure!

@zackw  For `char`, `u8` and `str`, I believe the implementations can simply be moved into the inherent impl blocks:

- [u8](https://github.com/rust-lang/rust/blob/1.15.0/src/libcore/num/mod.rs#L2285)
- [char](https://github.com/rust-lang/rust/blob/1.15.0/src/libstd_unicode/char.rs#L134)
- [str](https://github.com/rust-lang/rust/blob/1.15.0/src/libcollections/str.rs#L174)

For `[u8]`, we'll have add a new `u8_slice` lang time to allow an inherent `impl [u8]` alongside the generic [`impl<T> [T]`](https://github.com/rust-lang/rust/blob/1.15.0/src/libcollections/slice.rs#L169) block. The magic for that is [around here](https://github.com/rust-lang/rust/blob/1.15.0/src/libcollections/slice.rs#L169) but I'm not sure if the change will be straightforward. I can look into that part.

**Comment by sfackler** [comment]
Might be simpler to just stick it in the `imp<T>` block with a `where T: u8` bound on the methods.

**Comment by japaric** [comment]
>  with a where T: u8 bound on the methods.

But `u8` is a type, not a trait. Or are you suggesting a "double dispatch" approach?

**Comment by sfackler** [comment]
Oh duh, never mind me.

File: src/libstd/ascii.rs [hunk 1]
```diff
@@ -184,6 +184,368 @@ pub trait AsciiExt {
...
+    /// assert!(!lf.is_ascii_punctuation());
+    /// assert!(!esc.is_ascii_punctuation());
+    /// ```
+    #[unstable(feature = "ascii_ctype", issue = "39658")]
+    fn is_ascii_punctuation(&self) -> bool { unimplemented!(); }
+
+    /// Checks if the value is an ASCII graphic character:
+    /// U+0021 '@' ... U+007E '~'.
+    /// For strings, true if all characters in the string are
+    /// ASCII punctuation.
```
Comment (SimonSapin) [inline src/libstd/ascii.rs]: s/punctuation/graphic/ here.

**Comment by SimonSapin** [comment]
When doing this, please make sure that `[u8]` methods show up in rustdoc output.

**Pull request merged**